File: packages/openmemory-js/src/sources/notion.ts
/**
 * Notion Source Connector for OpenMemory.
 * Ingests pages and databases from Notion workspaces.
 * Requires: @notionhq/client
 * Environment: NOTION_API_KEY (optional fallback)
 */

import type { Client } from "@notionhq/client";

import { env } from "../core/cfg";
import { logger } from "../utils/logger";
import {
    BaseSource,
    SourceConfigError,
    SourceContent,
    SourceFetchError,
    SourceItem,
} from "./base";

interface NotionPage {
    id: string;
    url: string;
    last_edited_time: string;
    properties: Record<string, unknown>;
}

interface NotionBlock {
    type: string;
    [key: string]: unknown;
}

interface NotionCreds {
    apiKey?: string;
}

interface NotionFilters {
    databaseId?: string;
}

/**
 * Notion Source Connector.
 * Ingests pages and databases from Notion.
 */
export class NotionSource extends BaseSource<NotionCreds, NotionFilters> {
    override name = "notion";
    private client: Client | null = null;

    /**
     * Authenticates with Notion API.
     * Uses `process.env.NOTION_API_KEY` or passed `apiKey`.
     */
    async _connect(creds: NotionCreds): Promise<boolean> {
        let ClientConstructor: typeof Client;
        try {
            ClientConstructor = await import("@notionhq/client").then(
                (m) => m.Client,
            );
        } catch {
            throw new SourceConfigError(
                "missing deps: npm install @notionhq/client",
                this.name,
            );
        }

        // Security: BaseSource.connect has already hydrated creds from Persisted Config
        // Fallback to env.notionApiKey if no apiKey provided at all
        const apiKey = creds.apiKey || env.notionApiKey;

        if (!apiKey) {
            throw new SourceConfigError(
                "Notion API key is required (provide in Dashboard or OM_NOTION_API_KEY)",
                this.name,
            );
        }

        this.client = new ClientConstructor({ auth: apiKey });
        return true;
    }

    private extractTitle(page: NotionPage): string {
        const props = page.properties || {};
        for (const prop of Object.values(props)) {
            const p = prop as {
                type: string;
                title?: { plain_text: string }[];
            };
            if (
                p &&
                typeof p === "object" &&
                p.type === "title" &&
                Array.isArray(p.title) &&
                p.title[0]
            ) {
                return p.title[0].plain_text || "";
            }
        }
        return "";
    }

    /**
     * Lists Pages or Database Entries.
     * If `databaseId` is provided, queries that specific database.
     * Otherwise, searches for all accessible pages (Search API).
     *
     * @param filters - Filter constraints (databaseId).
     */
    async _listItems(filters: NotionFilters): Promise<SourceItem[]> {
        if (!this.client)
            throw new SourceConfigError("not connected", this.name);

        const results: SourceItem[] = [];

        if (filters.databaseId) {
            let hasMore = true;
            let startCursor: string | undefined;

            while (hasMore) {
                const resp = await (this.client.databases as any).query({
                    database_id: filters.databaseId,
                    start_cursor: startCursor,
                });

                for (const page of resp.results) {
                    if ("properties" in page) {
                        const p = page as unknown as NotionPage;
                        results.push({
                            id: page.id,
                            name: this.extractTitle(p) || "Untitled",
                            type: "page",
                            url: p.url || "",
                            lastEdited: p.last_edited_time,
                        });
                    }
                }

                if (results.length >= 1000) {
                    logger.warn(`[notion] Hit hard limit of 1000 items`);
                    break;
                }

                hasMore = resp.has_more;
                startCursor = resp.next_cursor || undefined;
                await this.rateLimiter.acquire();
            }
        } else {
            let hasMore = true;
            let startCursor: string | undefined;

            while (hasMore) {
                const resp = await this.client.search({
                    filter: { property: "object", value: "page" },
                    start_cursor: startCursor,
                });

                for (const page of resp.results) {
                    if ("properties" in page) {
                        const p = page as unknown as NotionPage;
                        results.push({
                            id: page.id,
                            name: this.extractTitle(p) || "Untitled",
                            type: "page",
                            url: p.url || "",
                            lastEdited: p.last_edited_time,
                        });
                    }
                }

                if (results.length >= 1000) {
                    logger.warn(`[notion] Hit hard limit of 1000 items (search)`);
                    break;
                }

                hasMore = resp.has_more;
                startCursor = resp.next_cursor || undefined;
                await this.rateLimiter.acquire();
            }
        }

        return results;
    }

    private blockToText(block: NotionBlock): string {
        const texts: string[] = [];
        const type = block.type;

        const textBlocks = [
            "paragraph",
            "heading_1",
            "heading_2",
            "heading_3",
            "bulleted_list_item",
            "numbered_list_item",
            "quote",
            "callout",
        ];

        if (textBlocks.includes(type)) {
            const content = block[type as keyof NotionBlock] as {
                rich_text?: { plain_text: string }[];
            };
            const richText = content?.rich_text || [];
            for (const rt of richText) {
                texts.push(rt.plain_text || "");
            }
        } else if (type === "code") {
            const content = block.code as {
                rich_text?: { plain_text: string }[];
                language?: string;
            };
            const richText = content?.rich_text || [];
            const lang = content?.language || "";
            const code = richText
                .map((rt: { plain_text: string }) => rt.plain_text || "")
                .join("");
            texts.push(`\`\`\`${lang}\n${code}\n\`\`\``);
        } else if (type === "to_do") {
            const content = block.to_do as {
                checked?: boolean;
                rich_text?: { plain_text: string }[];
            };
            const checked = content?.checked || false;
            const richText = content?.rich_text || [];
            const prefix = checked ? "[x] " : "[ ] ";
            texts.push(
                prefix +
                richText
                    .map(
                        (rt: { plain_text: string }) => rt.plain_text || "",
                    )
                    .join(""),
            );
        }

        return texts.join("");
    }

    /**
     * Fetches a Notion Page and converts its child blocks to Markdown.
     *
     * @param itemId - The UUID of the page.
     */
    async _fetchItem(itemId: string): Promise<SourceContent> {
        if (!this.client)
            throw new SourceConfigError("not connected", this.name);

        let page: unknown;
        try {
            page = await this.client.pages.retrieve({ page_id: itemId });
        } catch (e: unknown) {
            const msg = e instanceof Error ? e.message : String(e);
            throw new SourceFetchError(
                msg,
                this.name,
                e instanceof Error ? e : undefined,
            );
        }

        if (!page || typeof page !== "object" || !("properties" in page)) {
            throw new Error(`Item ${itemId} is not a valid page`);
        }

        const p = page as unknown as NotionPage;
        const title = this.extractTitle(p);
        const url = p.url || "";

        const blocks: NotionBlock[] = [];
        let hasMore = true;
        let startCursor: string | undefined;

        while (hasMore) {
            const resp = await this.client.blocks.children.list({
                block_id: itemId,
                start_cursor: startCursor,
            });
            blocks.push(...(resp.results as unknown as NotionBlock[]));
            hasMore = resp.has_more;
            startCursor = resp.next_cursor || undefined;
            await this.rateLimiter.acquire();
        }

        const textParts = title ? [`# ${title}`] : [];

        for (const block of blocks) {
            const txt = this.blockToText(block);
            if (txt.trim()) textParts.push(txt);
        }

        const text = textParts.join("\n\n");

        return {
            id: itemId,
            name: title || "Untitled",
            type: "notion_page",
            text,
            data: text,
            metadata: {
                source: "notion",
                pageId: itemId,
                url,
                blockCount: blocks.length,
            },
        };
    }
}

File: packages/openmemory-py/src/openmemory/connectors/notion.py
"""
notion connector for openmemory
requires: notion-client
env vars: NOTION_API_KEY
"""
from typing import List, Dict, Optional, Any
import os
import logging
from .base import BaseConnector, SourceItem, SourceContent, SourceAuthError

logger = logging.getLogger("openmemory.connectors.notion")

class NotionConnector(BaseConnector):
    """connector for notion pages and databases"""

    name = "notion"

    def __init__(self, user_id: Optional[str] = None):
        super().__init__(user_id)
        self.client = None

    async def _connect(self, **creds) -> bool:  # type: ignore[misc]
        """authenticate with notion api using AsyncClient"""
        try:
            from notion_client import AsyncClient  # type: ignore[import]
        except ImportError:
            raise ImportError("pip install notion-client")

        api_key = creds.get("api_key") or os.environ.get("NOTION_API_KEY")

        if not api_key:
            raise ValueError("no notion api key provided")

        self.client = AsyncClient(auth=api_key)
        self._connected = True
        return True

    async def _list_items(self, **filters) -> List[SourceItem]:
        """list pages from notion"""
        database_id = filters.get("database_id")

        if not self.client:
            raise SourceAuthError("Not connected", self.name)

        results = []  # type: ignore[unreachable]

        if database_id:
            has_more = True
            start_cursor = None

            while has_more:
                if not self.client or not hasattr(self.client, "databases"):
                    break

                resp = await self.client.databases.query(  # type: ignore[union-attr]
                    database_id=database_id, start_cursor=start_cursor
                )

                for page in resp.get("results", []):
                    title = self._extract_title(page)

                    results.append(SourceItem(
                        id=page["id"],
                        name=title or "Untitled",
                        type="page",
                        path=f"page/{page['id']}",
                        size=0,
                        metadata={  # type: ignore[arg-type]  # type: ignore[arg-type]
                            "url": page.get("url", ""),
                            "last_edited": page.get("last_edited_time")
                        }
                    ))

                has_more = resp.get("has_more", False)
                start_cursor = resp.get("next_cursor")
        else:
            resp = await self.client.search(filter={"property": "object", "value": "page"})

            for page in resp.get("results", []):
                title = self._extract_title(page)

                results.append(SourceItem(
                    id=page["id"],
                    name=title or "Untitled",
                    type="page",
                    path=f"page/{page['id']}",
                    size=0,
                    metadata={  # type: ignore[arg-type]  # type: ignore[arg-type]
                        "url": page.get("url", ""),
                        "last_edited": page.get("last_edited_time")
                    }
                ))

        return results

    def _extract_title(self, page: Dict) -> str:
        props = page.get("properties", {})
        for prop in props.values():
            if prop.get("type") == "title":
                titles = prop.get("title", [])
                if titles:
                    return titles[0].get("plain_text", "")
        return ""

    async def _fetch_item(self, item_id: str) -> SourceContent:
        """
        Fetch page content including all nested blocks as text.

        Args:
            item_id: The ID of the Notion page to fetch.

        Returns:
            SourceContent containing full page text.
        """
        if not self.client:
            raise SourceAuthError("Not connected", self.name)

        # get page metadata
        page = await self.client.pages.retrieve(page_id=item_id)
        title = self._extract_title(page)

        # recursively get all blocks
        blocks = await self._get_all_blocks_recursive(item_id)

        text_parts = [f"# {title}"] if title else []
        for block in blocks:
            txt = self._block_to_text(block)
            if txt.strip():
                text_parts.append(txt)

        text = "\n\n".join(text_parts)

        return SourceContent(
            id=item_id,
            name=title or "Untitled",
            type="notion_page",
            text=text,
            data=text,
            metadata={  # type: ignore[arg-type]  # type: ignore[arg-type]
                "source": "notion",
                "page_id": item_id,
                "url": page.get("url", ""),
                "block_count": len(blocks)
            }
        )

    async def _get_all_blocks_recursive(self, parent_id: str, depth: int = 0) -> List[Dict]:
        """Fetch all blocks for a given parent, including children recursively."""
        if not self.client:
            raise SourceAuthError("Not connected", self.name)
        if depth > 10: return [] # Limit depth to prevent infinite recursion or extreme latency

        blocks = []
        has_more = True
        start_cursor = None

        while has_more:
            resp = await self.client.blocks.children.list(
                block_id=parent_id,
                start_cursor=start_cursor
            )
            results = resp.get("results", [])
            for block in results:
                blocks.append(block)
                if (block or {}).get("has_children"):  # type: ignore[union-attr]
                    children = await self._get_all_blocks_recursive(
                        block.get("id", ""), depth + 1
                    )
                    blocks.extend(children)

            has_more = resp.get("has_more", False)
            start_cursor = resp.get("next_cursor")
            if len(blocks) > 1000: break # Higher safety limit for recursive

        return blocks

    def _block_to_text(self, block: Dict) -> str:
        """Convert a single Notion block to a text string."""
        texts = []
        block_type = block.get("type", "")

        if block_type in ["paragraph", "heading_1", "heading_2", "heading_3",
                          "bulleted_list_item", "numbered_list_item", "quote", "callout"]:
            rich_text = block.get(block_type, {}).get("rich_text", [])
            for rt in rich_text:
                texts.append(rt.get("plain_text", ""))

        elif block_type == "code":
            rich_text = block.get("code", {}).get("rich_text", [])
            for rt in rich_text:
                texts.append(rt.get("plain_text", ""))

        elif block_type == "to_do":
            checked = block.get("to_do", {}).get("checked", False)
            rich_text = block.get("to_do", {}).get("rich_text", [])
            prefix = "[x] " if checked else "[ ] "
            for rt in rich_text:
                texts.append(prefix + rt.get("plain_text", ""))

        elif block_type == "toggle":
            rich_text = block.get("toggle", {}).get("rich_text", [])
            for rt in rich_text:
                texts.append(rt.get("plain_text", ""))

        return "".join(texts)
